### PR TITLE
Commander: Separate out manual control setpoint processing

### PR DESCRIPTION
**Pull request by MaEtUgR**
**Describe problem/solution in this pull request**
New `ManualControl` class for handling `manual_control_setpoint` (and later switches) holding subscription, parameters and state necessary. It handles in this pr RC loss, RC override, RC arming/disarming.

- RC override: option to ignore throttle stick
- RC override: based on change (not centered stick)
- RC arming: based on Hysteresis instead of "loop ticker"
- RC arming: simplify the logic
- manual_control_setpoint message 

Note: This is the first stab at it and I'd like to do incremental steps to not create a non-understandable monster pr.

**Test data / coverage**
I tested the different cases during development but will do a final pass testing all the things with this state.

**Additional context**
Follow up after being in the loop from reviewing #16266 and doing work on RC override.

### CODE REVIEW

**Comment by MaEtUgR** [comment]
As a response to @dagar 's feedback:

> looks like a good direction
> my only initial suggestion/question is if we can get to the point where it's more event driven so that as much code as possible is only ever executed as a result of manual_control_setpoint updating

I prepared to gate the processing on `manual_control_setpoint` updates:
https://github.com/PX4/PX4-Autopilot/compare/throttle-stick-override-upstream...rc-handling
Testing necessary but then we can add it as well.

**Comment by MaEtUgR** [comment]
Arming/disarming not working as expected on a vehicle without props. Debugging.

**Comment by MaEtUgR** [comment]
I found the issue: ebbc098
Ready for review. I'll keep on testing and will only put bug fixes on top if necessary. Arming, disarming using all methods works.

**Comment by MaEtUgR** [comment]
Found another issue: Real-world RC seems to bounce RC loss, not reproducible on master on in SITL.

**Comment by MaEtUgR** [comment]
Fixed, squashed and rebased.

File: src/modules/commander/ManualControl.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+	OVERRIDE_IGNORE_THROTTLE_BIT = (1 << 2)
+};
+
+bool ManualControl::update()
+{
+	bool ret = false;
+
+	_rc_available = _rc_allowed
+			&& _last_manual_control_setpoint.timestamp != 0
+			&& (hrt_elapsed_time(&_last_manual_control_setpoint.timestamp) < (_param_com_rc_loss_t.get() * 1_s));
```
Comment (MaEtUgR) [inline src/modules/commander/ManualControl.cpp:51]: Using `_last_manual_control_setpoint` here is not good. It artificially shortens the timeout.

Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp]: 7789288

File: src/modules/commander/Commander.cpp [hunk 2]
```diff
@@ -426,18 +426,15 @@ transition_result_t Commander::arm(arm_disarm_reason_t calling_reason, bool run_
 
 	if (run_preflight_checks) {
 		if (_vehicle_control_mode.flag_control_manual_enabled) {
-			const bool throttle_above_low = (_manual_control_setpoint.z > 0.1f);
-			const bool throttle_above_center = (_manual_control_setpoint.z > 0.6f);
-
-			if (_vehicle_control_mode.flag_control_climb_rate_enabled && throttle_above_center) {
-				mavlink_log_critical(&_mavlink_log_pub, "Arming denied! Throttle not centered");
+			if (_vehicle_control_mode.flag_control_climb_rate_enabled && _manual_control.isThrottleAboveCenter()) {
+				mavlink_log_critical(&_mavlink_log_pub, "Arming denied because throttle above center");
```
Comment (bresch) [inline src/modules/commander/Commander.cpp:430]:
```suggestion
				mavlink_log_critical(&_mavlink_log_pub, "Arming denied: throttle above center");
```
Saving a few chars without loosing clarity.

File: src/modules/commander/Commander.cpp [hunk 3]
```diff
@@ -426,18 +426,15 @@ transition_result_t Commander::arm(arm_disarm_reason_t calling_reason, bool run_
...
 				tune_negative(true);
 				return TRANSITION_DENIED;
 
 			}
 
-			if (!_vehicle_control_mode.flag_control_climb_rate_enabled && throttle_above_low) {
-				mavlink_log_critical(&_mavlink_log_pub, "Arming denied! Throttle not zero");
+			if (!_vehicle_control_mode.flag_control_climb_rate_enabled && !_manual_control.isThrottleLow()) {
+				mavlink_log_critical(&_mavlink_log_pub, "Arming denied because of high throttle");
```
Comment (bresch) [inline src/modules/commander/Commander.cpp:437]:
```suggestion
				mavlink_log_critical(&_mavlink_log_pub, "Arming denied: high throttle");
```

File: src/modules/commander/ManualControl.cpp [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+				const vehicle_status_s &vehicle_status,
+				manual_control_switches_s &manual_control_switches, const bool landed)
+{
+	bool ret = false;
+
+	// no switch or button is mapped
+	const bool use_stick = manual_control_switches.arm_switch == manual_control_switches_s::SWITCH_POS_NONE;
+	// arm switch mapped and "switch is button" configured
+	const bool use_button = !use_stick && _param_com_arm_swisbtn.get();
+	// arm switch mapped and "switch_is_button" configured
```
Comment (bresch) [inline src/modules/commander/ManualControl.cpp:108]: copypasta!

File: src/modules/commander/ManualControl.cpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+bool ManualControl::wantsArm(const vehicle_control_mode_s &vehicle_control_mode, const vehicle_status_s &vehicle_status,
+			     manual_control_switches_s &manual_control_switches, const bool landed)
+{
+	bool ret = false;
+
+	// no switch or button is mapped
+	const bool use_stick = manual_control_switches.arm_switch == manual_control_switches_s::SWITCH_POS_NONE;
+	// arm switch mapped and "switch is button" configured
+	const bool use_button = !use_stick && _param_com_arm_swisbtn.get();
+	// arm switch mapped and "switch_is_button" configured
```
Comment (bresch) [inline src/modules/commander/ManualControl.cpp:154]:
```suggestion
	// arm switch mapped and "switch_is_button" not configured
```
not sure if the comment is useful

**Review by bresch — CHANGES_REQUESTED** [review]

File: src/modules/commander/ManualControl.cpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "ManualControl.hpp"
+#include <drivers/drv_hrt.h>
+
+using namespace time_literals;
+
+enum OverrideBits {
```
Comment (julianoes) [inline src/modules/commander/ManualControl.cpp:39]: Why not `enum class`?

File: src/modules/commander/ManualControl.cpp [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+	OVERRIDE_AUTO_MODE_BIT = (1 << 0),
+	OVERRIDE_OFFBOARD_MODE_BIT = (1 << 1),
+	OVERRIDE_IGNORE_THROTTLE_BIT = (1 << 2)
+};
+
+bool ManualControl::update()
+{
+	bool ret = false;
```
Comment (julianoes) [inline src/modules/commander/ManualControl.cpp:47]:
```suggestion
	bool updated = false;
```

File: src/modules/commander/ManualControl.cpp [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+	if (_manual_control_setpoint_sub.updated()) {
+		manual_control_setpoint_s manual_control_setpoint;
+
+		if (_manual_control_setpoint_sub.copy(&manual_control_setpoint)) {
+			process(manual_control_setpoint);
+		}
+
+		ret = true;
```
Comment (julianoes) [inline src/modules/commander/ManualControl.cpp:56]:
```suggestion
		updated = true;
```

File: src/modules/commander/ManualControl.hpp [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+	ManualControl(ModuleParams *parent) : ModuleParams(parent) {};
+	~ManualControl() override = default;
+
+	void setRCAllowed(const bool rc_allowed) { _rc_allowed = rc_allowed; }
+
+	/**
+	 * Check for manual control input changes and process them
+	 * @return true if there was new data
+	 */
+	bool update();
```
Comment (julianoes) [inline src/modules/commander/ManualControl.hpp:65]: I'm not super happy with the name `update()` for something that also returns something but I also don't have a better suggestion, unless maybe `couldUpdate()` or `tryUpdate()`?

File: src/modules/commander/ManualControl.hpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+	bool isRCAvailable() { return _rc_available; }
```
Comment (julianoes) [inline src/modules/commander/ManualControl.hpp:66]: Most of these can be marked `const`:

```suggestion
	bool isRCAvailable() const { return _rc_available; }
```

File: src/modules/commander/ManualControl.cpp [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+	}
+
+	_rc_available = _rc_allowed
+			&& _manual_control_setpoint.timestamp != 0
+			&& (hrt_elapsed_time(&_manual_control_setpoint.timestamp) < (_param_com_rc_loss_t.get() * 1_s));
+
+	return ret && _rc_available;
+}
+
+void ManualControl::process(manual_control_setpoint_s &manual_control_setpoint)
```
Comment (julianoes) [inline src/modules/commander/ManualControl.cpp:66]:
```suggestion
void ManualControl::process(const manual_control_setpoint_s &manual_control_setpoint)
```

File: src/modules/commander/ManualControl.cpp [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+					    && vehicle_control_mode.flag_control_offboard_enabled;
+
+	if (_rc_available && (override_auto_mode || override_offboard_mode)) {
+		const float minimum_stick_change = .01f * _param_com_rc_stick_ov.get();
+
+		const bool rpy_moved = (fabsf(_manual_control_setpoint.x - _last_manual_control_setpoint.x) > minimum_stick_change)
+				       || (fabsf(_manual_control_setpoint.y - _last_manual_control_setpoint.y) > minimum_stick_change)
+				       || (fabsf(_manual_control_setpoint.r - _last_manual_control_setpoint.r) > minimum_stick_change);
+		const bool throttle_moved =
+			(fabsf(_manual_control_setpoint.z - _last_manual_control_setpoint.z) * 2.f > minimum_stick_change);
```
Comment (julianoes) [inline src/modules/commander/ManualControl.cpp:87]: Can you add a comment why there is a factor of 2 here?

File: src/modules/commander/ManualControl.cpp [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+	} else if (!arm_button_pressed) {
+
+		_stick_disarm_hysteresis.set_state_and_update(false, hrt_absolute_time());
+	}
+
+	return ret;
+}
+
+bool ManualControl::wantsArm(const vehicle_control_mode_s &vehicle_control_mode, const vehicle_status_s &vehicle_status,
+			     manual_control_switches_s &manual_control_switches, const bool landed)
```
Comment (julianoes) [inline src/modules/commander/ManualControl.cpp:146]:
```suggestion
bool ManualControl::wantsArm(const vehicle_control_mode_s &vehicle_control_mode, const vehicle_status_s &vehicle_status,
			     const manual_control_switches_s &manual_control_switches, const bool landed)
```

File: src/modules/commander/ManualControl.cpp [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+{
+	bool ret = false;
+
+	// no switch or button is mapped
+	const bool use_stick = manual_control_switches.arm_switch == manual_control_switches_s::SWITCH_POS_NONE;
```
Comment (julianoes) [inline src/modules/commander/ManualControl.cpp:151]: All this logic below is pretty confusing. I wonder if it would be clearer by removing the comments and then think about the naming again.

E.g. here:

```suggestion
	const bool no_switch_mapped = manual_control_switches.arm_switch == manual_control_switches_s::SWITCH_POS_NONE;
```

**Review by julianoes — COMMENTED** [review]
Couple of comments but overall that's a very nice cleanup, exactly the sort of thing we should do step by step.

File: src/modules/commander/ManualControl.hpp [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+	bool wantsOverride(const vehicle_control_mode_s &vehicle_control_mode);
+	bool wantsDisarm(const vehicle_control_mode_s &vehicle_control_mode, const vehicle_status_s &vehicle_status,
+			 manual_control_switches_s &manual_control_switches, const bool landed);
+	bool wantsArm(const vehicle_control_mode_s &vehicle_control_mode, const vehicle_status_s &vehicle_status,
+		      manual_control_switches_s &manual_control_switches, const bool landed);
+	bool isThrottleLow() { return _last_manual_control_setpoint.z < 0.1f; }
+	bool isThrottleAboveCenter() { return _last_manual_control_setpoint.z > 0.6f; }
+	hrt_abstime getLastRCTimestamp() { return _last_manual_control_setpoint.timestamp; }
```
Comment (JonasVautherin) [inline src/modules/commander/ManualControl.hpp:74]: Just a nitpick (no need to change it here, but I want to share my view because I just discussed it with @julianoes). For a long time, I never knew if I should write `GPS` vs `Gps`, or `getLastRcTimestamp` vs `getLastRcTimestamp`. Working on an auto-generation system, I realized that the latter was always better.

I see it like this: `getLastRcTimestamp` is exactly equivalent to `get_last_rc_timestamp`, but `getLastRCTimestamp` requires vocabulary knowledge (is it `get_last_r_c_timestamp`, or maybe `get_last_r_ctimestamp`?). What if one day you want to parse the name?

And to give a real example: the Swift version of protobuf written by Apple tries to go the aesthetic way, and uses `HTTP`, `ID`, and exactly 2 others I don't remember. But that requires a vocabulary of acronyms, and since they don't know `RC`, it has to be `Rc`. Worse, they know `HTTP`, but they use `Ftp` (how consistent, eh?). Even worse: they have extra code handling those 4 exceptions (they need to go change `Http` to `HTTP` and `Id` to `ID` everywhere), and they have extra unit tests for that.

TL;DR: Camelcase says that each word starts with a capital letter. `GPS` is a word, so in camelcase, it becomes `Gps`.

File: src/modules/commander/ManualControl.cpp (hunk 6 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp:39]: No reason, good suggestion. It was added like that:
https://github.com/PX4/PX4-Autopilot/pull/13618/files#diff-33729b6612e70db2803e6b18b64a47c71e70eaf1d6b69fd233a34b3b0408a806R261
I'm changing it 👍

Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp:39]: Oh wait there is a reason, It gets used in combination with a bitwise and not direct comparison. Would you define the bitwise and operator in the class or how can it be done easily? Holding back with the change because of lack of good ideas to solve this.

File: src/modules/commander/ManualControl.cpp (hunk 4 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp:108]: At least consistent 😄

File: src/modules/commander/ManualControl.cpp (hunk 5 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp:154]: Probably they just helped me develop. The variable names are clear and the conditions aren't crazy. Removed them.

File: src/modules/commander/ManualControl.cpp (hunk 6 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp:39]: It gets used in combination with a "bitwise and" not in a direct comparison. Would you define the bitwise and operator in the class or how can it be done more easily?

That's probably also why it was added in this way:
https://github.com/PX4/PX4-Autopilot/pull/13618/files#diff-33729b6612e70db2803e6b18b64a47c71e70eaf1d6b69fd233a34b3b0408a806R261

File: src/modules/commander/ManualControl.hpp (hunk 9 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.hpp:65]: I thought about this because I also don't like if everything is called update() for e.g. text search reasons. But updated() is wrong because it actually does something and I couldn't come up with something better. Maybe checkUpdateAndProcess()? Seems clunky and doesn't really help does it?

File: src/modules/commander/ManualControl.cpp (hunk 12 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp:87]: Because https://github.com/PX4/PX4-Autopilot/pull/15949 was not merged 😞  I'll rebase it, make it mergeable again, and then push harder.

Also, I added a comment.

File: src/modules/commander/ManualControl.cpp (hunk 14 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp:151]: The reason I did it like this is an in-between step where I would consider it more clear what's going on than before and it enables to separate the different arming methods completely with less combined confusion and a separately reported reason per method.

File: src/modules/commander/ManualControl.hpp (hunk 15 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.hpp:74]: You are right! I honestly didn't think that far and just wrote it how I considered it most readable but I don't see any disadvantage in terms of understanding with pure camel case but an advantage in consistency 👍

Comment (MaEtUgR) [reply src/modules/commander/ManualControl.hpp]: d909125

File: src/modules/commander/ManualControl.cpp (hunk 13 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp]: d909125

File: src/modules/commander/ManualControl.cpp (hunk 12 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp:88]: d909125

File: src/modules/commander/ManualControl.cpp (hunk 11 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp]: d909125

File: src/modules/commander/ManualControl.hpp (hunk 10 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.hpp]: I also added the const qualifier to the other getters.
d909125

File: src/modules/commander/ManualControl.cpp (hunk 8 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp]: Much better! 🤦  d909125

File: src/modules/commander/ManualControl.cpp (hunk 7 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp]: d909125

File: src/modules/commander/ManualControl.cpp (hunk 5 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp]: d909125

File: src/modules/commander/ManualControl.cpp (hunk 4 above)
Comment (MaEtUgR) [reply src/modules/commander/ManualControl.cpp]: d909125

File: src/modules/commander/Commander.cpp (hunk 2 above)
Comment (MaEtUgR) [reply src/modules/commander/Commander.cpp]: That's right, clear != long d909125

File: src/modules/commander/Commander.cpp (hunk 3 above)
Comment (MaEtUgR) [reply src/modules/commander/Commander.cpp]: d909125

**Review by julianoes — APPROVED** [review]
Thanks for the fixes @MaEtUgR!

**Comment by MaEtUgR** [comment]
Thank you for the reviews @julianoes and @bresch and @JonasVautherin, that really helped a lot. I'll follow up with further improvements like freeing commander from the rc_found_once logic, doing all the processing in the process() function, separating arming methods.
